drivers/SmartThings/zigbee-switch/src/aqara/init.lua

```diff
@@ -70,140 +116,251 @@ local function is_aqara_products(opts, driver, device)
+  local version = tonumber(value.value)
+  device:set_field(APPLICATION_VERSION, version, { persist = true })
+end
+
+local function round(num)
+  local mult = 10
+  return math.floor(num * mult + 0.5) / mult
+end
+
+local function energy_meter_power_consumption_report(device, raw_value)
+  -- report interval
+  local current_time = os.time()
+  local last_time = device:get_field(LAST_REPORT_TIME) or 0
+  local next_time = last_time + 60 * 15 -- 15 mins, the minimum interval allowed between reports
+  if current_time < next_time then
+    return
+  end
+  device:set_field(LAST_REPORT_TIME, current_time, { persist = true })
+
+  -- energy meter
+  device:emit_event(capabilities.energyMeter.energy({ value = raw_value, unit = "Wh" }))
+
+  -- power consumption report
   local delta_energy = 0.0
   local current_power_consumption = device:get_latest_state("main", capabilities.powerConsumptionReport.ID,
     capabilities.powerConsumptionReport.powerConsumption.NAME)
   if current_power_consumption ~= nil then
     delta_energy = math.max(raw_value - current_power_consumption.energy, 0.0)
   end
   device:emit_event(capabilities.powerConsumptionReport.powerConsumption({ energy = raw_value, deltaEnergy = delta_energy })) -- the unit of these values should be 'Wh'
+
 end
 
-local function application_version_handler(driver, device, value, zb_rx)
-  local version = tonumber(value.value)
-  device:set_field(APPLICATION_VERSION, version, { persist = true })
+local function energy_meter_handler(driver, device, value, zb_rx)
+  if device:get_model() == "lumi.plug.maeu01" then
+    local version = device:get_field(APPLICATION_VERSION) or 0
+    if version == 32 then
+      -- not allowed
+      return
+    end
+  end
+  local raw_value = value.value -- 'Wh'
+  energy_meter_power_consumption_report(device, raw_value)
 end
 
 local function power_meter_handler(driver, device, value, zb_rx)
+  if device:get_model() == "lumi.plug.maeu01" then
+    local version = device:get_field(APPLICATION_VERSION) or 0
+    if version == 32 then
+      -- not allowed
+      return
+    end
+  end
+
   local raw_value = value.value -- '10W'
   raw_value = raw_value / 10
   device:emit_event(capabilities.powerMeter.power({ value = raw_value, unit = "W" }))
 end
 
-local function energy_meter_handler(driver, device, value, zb_rx)
-  local raw_value = value.value -- 'Wh'
-  -- energyMeter
-  device:emit_event(capabilities.energyMeter.energy({ value = raw_value, unit = "Wh" }))
-  -- powerConsumptionReport
-  emit_power_consumption_report_event(device, { value = raw_value })
-end
-
-local function round(num)
-  local mult = 10
-  return math.floor(num * mult + 0.5) / mult
-end
-
 local function present_value_handler(driver, device, value, zb_rx)
+  if device:get_model() == "lumi.plug.maeu01" then
+    local version = device:get_field(APPLICATION_VERSION) or 0
+    if version == 41 then
+      -- not allowed
+      return
+    end
+  end
```

Comment:
This sort of thing should have been left as a sub-driver how it was. If you want one sub-driver that handles both of the version-related commands, that's fine, but in general, for handling of events, if you've included a model check, it should probably be a sub-driver.

Reply:
Moved plug's code to the sub-driver.